### PR TITLE
[Enhancement] counting semaphore used in bthread

### PR DESCRIPTION
**Pull request by sduzh**
This PR was splitted from another bigger pr https://github.com/StarRocks/starrocks/pull/33670

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
@kevincai @wyb PTAL

File: be/src/util/bthreads/semaphore.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+#include <bthread/processor.h> // cpu_relax()
+
+#include <chrono>
+#include <climits>
+
+#include "util/time.h"
+
+namespace starrocks::bthreads {
+
+// Modeled from std::counting_semaphore.
```
Comment (kevincai) [inline be/src/util/bthreads/semaphore.h:27]: inspired by ... :)

**Review by kevincai — APPROVED** [review]

File: be/src/util/bthreads/semaphore.h (hunk 1 above)
Comment (sduzh) [reply be/src/util/bthreads/semaphore.h]: Done

**Review by Astralidea — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by andyziye** [comment]
#### [BE Incremental Coverage Report]


:white_check_mark: **pass** : 0 / 0 (0%)

**Comment by andyziye** [comment]
#### [BE Incremental Coverage Report]


:white_check_mark: **pass** : 0 / 0 (0%)

**Comment by andyziye** [comment]
#### [BE Incremental Coverage Report]

:x: **fail** : 0 / 30 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/util/bthreads/semaphore.h|0|30|00.00%|[33, 34, 39, 40, 41, 42, 44, 45, 46, 59, 60, 61, 62, 63, 66, 67, 68, 69, 71, 86, 87, 88, 89, 90, 91, 92, 95, 96, 102, 103]|

**Comment by andyziye** [comment]
#### [BE Incremental Coverage Report]

:x: **fail** : 0 / 30 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/util/bthreads/semaphore.h|0|30|00.00%|[33, 34, 39, 40, 41, 42, 44, 45, 46, 59, 60, 61, 62, 63, 66, 67, 68, 69, 71, 86, 87, 88, 89, 90, 91, 92, 95, 96, 102, 103]|

**Pull request merged**